lib/kafka_ex/gen_consumer.ex

```diff
@@ -0,0 +1,612 @@
+  alias KafkaEx.Protocol.OffsetFetch.Response, as: OffsetFetchResponse
+  alias KafkaEx.Protocol.Offset.Response, as: OffsetResponse
+  alias KafkaEx.Protocol.Fetch.Response, as: FetchResponse
+  alias KafkaEx.Protocol.Fetch.Message
+
+  require Logger
+
+  @typedoc """
+  Option values used when starting a `KafkaEx.GenConsumer`.
+  """
+  @type option :: {:commit_interval, non_neg_integer}
+                | {:commit_threshold, non_neg_integer}
+
+  @typedoc """
+  Options used when starting a `KafkaEx.GenConsumer`.
+  """
+  @type options :: [option | GenServer.option]
+
+  @doc """
+  Invoked when the server is started. `start_link/5` will block until it
+  returns.
+
+  `topic` and `partition` are the arguments passed to `start_link/5`. They
+  identify the Kafka partition that the `KafkaEx.GenConsumer` will consume from.
+
+  Returning `{:ok, state}` will cause `start_link/5` to return `{:ok, pid}` and
+  the process to start consuming from its assigned partition. `state` becomes
+  the consumer's state.
+
+  Any other return value will cause the `start_link/5` to return `{:error,
+  error}` and the process to exit.
+  """
+  @callback init(topic :: binary, partition :: non_neg_integer) ::
+    {:ok, state :: term}
+
+  @doc """
+  Invoked for each message set consumed from a Kafka topic partition.
+
+  `message_set` is a message set fetched from a Kafka broker and `state` is the
+  current state of the `KafkaEx.GenConsumer`.
+
+  Returning `{:async_commit, new_state}` acknowledges `message` and continues
+  to consume from the Kafka queue with new state `new_state`. Acknowledged
+  messages will be auto-committed (possibly at a later time) based on the
+  `:commit_interval` and `:commit_threshold` options.
+
+  Returning `{:sync_commit, new_state}` commits `message` synchronously before
+  continuing to consume from the Kafka queue with new state `new_state`.
+  Committing a message synchronously means that no more messages will be
+  consumed until the message's offset is committed. `:sync_commit` should be
+  used sparingly, since committing every message synchronously would impact a
+  consumer's performance and could result in excessive network traffic.
+  """
+  @callback handle_message_set(message_set :: [Message.t], state :: term) ::
+    {:async_commit, new_state :: term} | {:sync_commit, new_state :: term}
+
+  @doc """
+  Invoked by `KafkaEx.GenConsumer.call/3`.
+
+  Note the default implementation will cause a `MatchError`.  If you want to
+  interact with your consumer, you must implement a handle_call function.
+  """
+  @callback handle_call(call :: term, from :: GenServer.from, state :: term)
+    :: {:reply, reply_value :: term, new_state :: term}
+
+  defmacro __using__(_opts) do
+    quote do
+      @behaviour KafkaEx.GenConsumer
+      alias KafkaEx.Protocol.Fetch.Message
+
+      def init(_topic, _partition) do
+        {:ok, nil}
+      end
+
+      def handle_call(_call, _from, _consumer_state) do
```

Comment:
I didn't see anywhere that it's explicitly called out in the documentation which calls should be handled when this is implemented - we call the `:partition` and `{:consumer_call, _}` forms from this module, so it seems like we should either spell that out, or implement a default callback.